### PR TITLE
Add Monoidal (->) tests

### PR DESCRIPTION
**Pull request by xplosunn**
Helps with #7 

I'm not sure if I understood the laws correctly and how they apply here. Any comments are appreciated.

### CODE REVIEW

**Comment by xplosunn** [comment]
Should I fix all hint-man comments?

File: test/Test/Prolens/Property.hs [hunk 1]
```diff
@@ -66,3 +67,33 @@ eqFun :: Fun Maybe Int Int -> Fun Maybe Int Int -> PropertyT IO ()
...
     unFun fun1 x === unFun fun2 x
+
+monoidalSpec :: Spec
+monoidalSpec = describe "Monoidal" $ do
+    describe "(->)" $ do
+        it "Identity: pappend pempty id ≡ id" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            pappend pempty f (x, y) === (x, f y)
```
Comment (vrom911) [inline test/Test/Prolens/Property.hs]: `id` can't be used, and you actually check another law here, which is correct 👍🏼 
However, we can make it even a bit more elegant and general, if we use the fact that it is `Strong`. We can check these properties instead:

```haskell
pappend f pempty ≡ first f
pappend pempty f ≡ second f
```

File: test/Test/Prolens/Property.hs [hunk 2]
```diff
@@ -66,3 +67,33 @@ eqFun :: Fun Maybe Int Int -> Fun Maybe Int Int -> PropertyT IO ()
...
+            y <- forAll genInt
+            pappend f pempty (x, y) === (f x, y)
+        it "Associativity (right)" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            g <- forAllWith (const "g") genFunction
+            h <- forAllWith (const "h") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            z <- forAll genInt
+            pappend f (pappend g h) (x, (y, z)) === (f x, (g y, h z))
```
Comment (vrom911) [inline test/Test/Prolens/Property.hs:98]: This looks right.
I think we could make it even more cooler, if we would check both associativity at once as well:
```haskell
pappend f (pappend g h) (x, (y, z))  === dimap (\(a, (b, c)) -> ((a, b), c))  (\((a, b), c) -> (a, (b, c)))(pappend (pappend f g) h (x, (y, z)) )
```
What do you think?

**Review by vrom911 — COMMENTED** [review]
Thank for your work @xplosunn ! I have added a few ideas in comments 🙂 

And you can ignore "duplication" warnings of hintman, no worries 👌🏼

File: test/Test/Prolens/Property.hs (hunk 2 above)
Comment (xplosunn) [reply test/Test/Prolens/Property.hs:98]: I think that line becomes quite long.

My counterproposal is to add two helper functions

```haskell
untupleRight :: (a, (b, c)) -> (a, b, c)
untupleRight (a, (b, c)) = (a, b, c)

untupleLeft :: ((a, b), c) -> (a, b, c)
untupleLeft ((a, b), c) = (a, b, c)
```

And the test would become

```haskell
it "Associativity" $ hedgehog $ do
    f <- forAllWith (const "f") genFunction
    g <- forAllWith (const "g") genFunction
    h <- forAllWith (const "h") genFunction
    x <- forAll genInt
    y <- forAll genInt
    z <- forAll genInt
    untupleRight (pappend f (pappend g h) (x, (y, z))) === untupleLeft (pappend (pappend f g) h ((x, y), z))
```

File: test/Test/Prolens/Property.hs [hunk 3]
```diff
@@ -66,3 +67,33 @@ eqFun :: Fun Maybe Int Int -> Fun Maybe Int Int -> PropertyT IO ()
...
     unFun fun1 x === unFun fun2 x
+
+monoidalSpec :: Spec
+monoidalSpec = describe "Monoidal" $ do
+    describe "(->)" $ do
+        it "Identity: pappend f pempty ≡ first f" $ hedgehog $ do
+            f <- forAllWith (const "f") genFunction
+            x <- forAll genInt
+            y <- forAll genInt
+            pappend f pempty (x, y) === (f x, y)
```
Comment (vrom911) [inline test/Test/Prolens/Property.hs]:
```suggestion
            pappend f pempty (x, y) === first f (x, y)
```

File: test/Test/Prolens/Property.hs (hunk 2 above)
Comment (vrom911) [reply test/Test/Prolens/Property.hs:98]: Sure, the helper functions are way to go. 
I don't think that `dimap` is too long to type, so I don't see why it is a problem..

Comment (xplosunn) [reply test/Test/Prolens/Property.hs:98]: To be honest I can't get it to compile with dimap. I'm always running into issues like 

```
Couldn't match type ‘Int’ with ‘((Int, b1), c0)’
      Expected type: (Int, (Int, Int))
                     -> (((a0, b0), Int), ((Int, b1), c0))
        Actual type: ((Int, Int), Int) -> ((Int, Int), Int)
```

Maybe it has something to do with the tuple type parameters?

Comment (vrom911) [reply test/Test/Prolens/Property.hs:98]: Okay, no worries. I would look into that separately 🙂

**Review by vrom911 — APPROVED** [review]

**Pull request merged**